consensus/polybft/consensus_runtime.go

```diff
@@ -354,6 +354,40 @@ func (c *consensusRuntime) FSM() error {
 	isEndOfSprint := c.isEndOfSprint(pendingBlockNumber)
 	isEndOfEpoch := c.isEndOfEpoch(pendingBlockNumber)
 
+	valSet, err := NewValidatorSet(epoch.Validators)
+	if err != nil {
+		return fmt.Errorf("cannot create validator set for fsm: %w", err)
+	}
+
+	if epoch.Number > 0 { // it must be at least epoc
+		iterationNumber := uint64(0)
+		currentHeader := parent
+		lastBlockOfPreviousEpoch := getEndEpochBlockNumber(epoch.Number-1, c.config.PolyBFTConfig.EpochSize)
+
+		for currentHeader.Number > lastBlockOfPreviousEpoch {
+			blockExtra, err := GetIbftExtra(currentHeader.ExtraData)
+			if err != nil {
+				return fmt.Errorf("cannot get ibft extra: %w", err)
+			}
+
+			iterationNumber += blockExtra.Round + 1 // because round 0 is one of the iterations
```

Comment:
I think we could optimize this, by keeping the `iterationNumber` in `epochMetadata` which we can update on `OnBlockInserted`, and only call this logic if node was dropped and restarted. We won't have a problem if `syncer` inserts the block first (since `OnBlockInserted` is called both from `syncer` and on `InsertBlock` from consensus), because if block was already inserted `OnBlockInserted` would not be called.

This is just a food for thought, we don't need to do it right away.

Reply:
yes, this needs refactoring